fedspendingtransparency/usaspending-api
# [DEV-6657] Check for snapshot before deleting ES index

**Description:**
Add logic to the `elasticsearch_indexer` that will check if any of the indexes that will be deleted are currently being used by a snapshot. If a snapshot is occurring then check every 90 seconds (up to 15 minutes) if it is safe to delete the index.

**Technical details:**
Query the Elasticsearch cluster for any ongoing snapshots. The index(es) that are to be deleted are then compared against those that are blocked by one or more snapshots. Additionally, when an exception occurred an error was occurring due to `logger.exception()` not having an `action` attribute. This is because `format_log()` needed to be present similar to uses of `logger.info()`.

If after 15 minutes the snapshot is still running then the job will complete. To allow the pipeline to move past this stage but still return an appropriate status of `UNSTABLE` when this occurs a new value was added to the config to track the `pipeline_exit_code`. In the event of an Index not being deleted due to a snapshot the `pipeline_exit_code` will be returned as `1`. Necessary changes will be made to the nightly pipeline to make sure that this flag determines if the pipeline is `UNSTABLE` or not without causing it to error out.

**Testing Notes:**
To test the happy path locally without the need to trigger a snapshot in a non-local environment the following steps were taken:
1. Access the Elasticsearch docker container and add `path.repo: ["/usr/share/elasticsearch/backup"]` to `/usr/share/elasticsearch/config/elasticsearch.yml`
2. Restart the cluster so that changes take effect
3. Create a snapshot repo for the Elasticsearch cluster:
> PUT /_snapshot/my_backup 
{
  "type": "fs",
  "settings": {
    "location": "/usr/share/elasticsearch/backup"
  }
}
4. Trigger the Elasticsearch Indexer with a breakpoint before the stage that checks for Snapshots
5. Once the job hits the necessary breakpoint start a snapshot:
> PUT /_snapshot/my_backup/snapshot_1
6. Move past the breakpoint to see the snapshot hold logic take effect

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. [x] API documentation updated
3. [ ] Necessary PR reviewers:
    - [ ] Domain Expert <OPTIONAL>
4. [x] Matview impact assessment completed (N/A)
5. [x] Frontend impact assessment completed (N/A)
6. [x] Data validation completed
7. [x] Appropriate Operations ticket(s) created
8. [x] Jira Ticket [DEV-6657](https://federal-spending-transparency.atlassian.net/browse/DEV-6657):
    - [x] Link to this Pull-Request
    - [x] Performance evaluation of affected (API | Script | Download)
    - [x] Before / After data comparison

**Area for explaining above N/A when needed:**
```
```
